### PR TITLE
add rephrased samples & training code & reproduce f1 score

### PR DESCRIPTION
**Pull request by andy-yang-1**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,4 +84,18 @@ python3 main.py --train_path ./data/train/CodeAlpaca-20k.jsonl \
 
 ## Dataset and Training Code
 
-_Coming Soon!_
\ No newline at end of file
+_Coming Soon!_
+
+## F1 Score
+
+Reproduce paper's Table 5 & 6
```
Comment (infwinston) [inline README.md:91]: Can we put a screenshot or markdown for Table 5 and 6 here?

File: f1score/humaneval/f1_emb.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            re_TP += 1
+        else:
+            re_FN += 1
+
+    print(re_TP)
+    print(re_FN)
+
+    re_f1 = compute_f1score(re_TP, FP, re_FN)
+
+    print(f"Rephrase {language} F1 score: {re_f1}")
```
Comment (infwinston) [inline f1score/humaneval/f1_emb.py]: add eol

File: f1score/humaneval/f1_emb.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def get_programs(r_path):
+    dataset = [json.loads(l) for l in open(r_path, "r")]
+    programs = [each["text"] for each in dataset]
+    return programs
+
+
+languages = ["python", "c", "js"]
+model = SentenceTransformer('multi-qa-MiniLM-L6-cos-v1')
```
Comment (infwinston) [inline f1score/humaneval/f1_emb.py]: double quote to be consistent

File: f1score/humaneval/f1_llm.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            re_TP += 1
+            continue
+        if detect_contamination(model, original_programs[i], rephrased_programs[i], instruct):
+            re_TP += 1
+        else:
+            re_FN += 1
+
+
+    re_f1 = compute_f1score(re_TP, FP, re_FN)
+    print(f"Rephrase {language} F1 score: {re_f1}")
```
Comment (infwinston) [inline f1score/humaneval/f1_llm.py]: ditto

File: f1score/mmlu/f1_emb.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+model = SentenceTransformer('multi-qa-MiniLM-L6-cos-v1')
+
+threshold = 0.5
+
+for subject in subjects:
+    original_questions = get_original_questions(f"data/rephrase/{subject}_test.csv")
+    rephrase_questions = get_rephrase_english_questions(f"data/rephrase/{subject}_test_rephrase_english_filtered_question.jsonl")
+    chinese_questions = get_chinese_questions(f"data/rephrase/{subject}_test_chinese_filtered_question.jsonl")
+
+    original_questions = original_questions[:100]
```
Comment (infwinston) [inline f1score/mmlu/f1_emb.py:67]: why `:100`?

File: f1score/mmlu/f1_llm.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    # Compute F1 score
+    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) != 0 else 0
+    
+    return f1
+
+
+def detect_contamination(model, question1, question2, instruct):
+
+    retries = 0
+    while retries < 30:
```
Comment (infwinston) [inline f1score/mmlu/f1_llm.py]: Add `MAX_RETRY=30` to make it adjustable?

File: f1score/mmlu/f1_llm.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            ch_FN += 1
+
+    re_f1 = compute_f1score(re_TP, FP, re_FN)
+    ch_f1 = compute_f1score(ch_TP, FP, ch_FN)
+    te_f1 = compute_f1score(100, FP, 0)
+
+
+    print(f"{subject} test set f1: {te_f1}")
+    print(f"{subject} rephrase f1: {re_f1}")
+    print(f"{subject} chinese f1: {ch_f1}")
```
Comment (infwinston) [inline f1score/mmlu/f1_llm.py]: ditto

File: finetune.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    --warmup_ratio 0.05 \
+    --lr_scheduler_type "cosine" \
+    --logging_steps 1 \
+    --save_strategy "steps" \
+    --save_steps 1000 \
+    --save_total_limit 20 \
+    --bf16 True \
+    --tf32 True \
+    --fsdp "full_shard auto_wrap" \
+    --fsdp_config train/fsdp_config.json
```
Comment (infwinston) [inline finetune.sh]: ditto

File: rephrase_tokenize.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    mmap_dataset.save(out_file)
+    print(f"Save to {out_file}")
+
+    if args.debug:
+        item = mmap_dataset[0]
+        token_ids = item["token_ids"]
+        is_ignore = item["is_ignore"]
+        padding_mask = item["padding_mask"]
+        token_ids[is_ignore] = tokenizer.unk_token_id
+        print(tokenizer.decode(token_ids))
```
Comment (infwinston) [inline rephrase_tokenize.py]: ditto for all files

**Review by infwinston — COMMENTED** [review]
thanks left some comments

File: f1score/mmlu/f1_emb.py (hunk 5 above)
Comment (andy-yang-1) [reply f1score/mmlu/f1_emb.py:67]: Because there are 200 pairs, 100 random pairs and 100 contamination pairs

**Comment by infwinston** [comment]
@andy-yang-1 merge?

**Pull request merged**